spec/form_spec.cr

```diff
@@ -15,6 +15,16 @@ end
 private class TaskForm < Task::BaseForm
 end
 
+private class UserCallbackForm < User::BaseForm
+  allow :id, :name, :age, :joined_at
+
+  def before_save
+    first = UserQuery.new.first
+    UserForm.new(first, name: "changed").save! # This doesn't use the same tx connection
```

Comment:
In order to trigger the transaction rollback, the nested `UserForm` needs to access the `UserCallbackForm`s transaction.

One possibility may be to store the transaction in the `User::BaseForm` as a class method, but then it's restricted to classes which isn't useful.

I imagine ideal way would be to get the context of `UserForm.new(...).save!`, meaning the `before_save` method then getting the `UserCallbackForm` from there, but I don't even know if that's possible.

Reply:
I believe you can also do this:

https://github.com/crystal-lang/crystal-db/blob/28b17b7dba820d7e4810493aaf0a99e368849ec6/spec/transaction_spec.cr#L68

I'm not entirely sure if it works with nested transactions, but I think it is worth a shot! If it works that would be incredible. If not, LMK and we'll try to think of something. My guess would be adding a `transaction` property on forms that is `nil` by default, but if it is set, it uses it instead of creating a new transaction

Reply:
So the idea is that raising `DB::Rollback.new` will bubble up to the next transaction? I'll try it out and see if I can make it work

Reply:
Yeah exactly. I'm fairly sure it should bubble *fingers crossed*

Reply:
I'm still getting the same error. `UserForm` is creating its own transaction and committing it before the execution comes back to `before_save`. 

From a quick glance at the code, it looks like Rails handles transactions by tracking them in `ActiveRecord::Base`. We can do something similar in BaseForm, but what happens in a multi threaded situation where a user spawns fibers that might interact with the db. Wouldn't the fibers all have access to the same `BaseForm.transaction`? (I'm still new to concurrency with fibers)